### DISCUSSION TITLE
issue saving many renders in a loop: "context leak detected"

### DISCUSSION THREAD

**Discussion by josephko91** [pyvista/pyvista#4274] [Q&A] [answered]
Hi. I'm trying to save many 2-d renders of a mesh with random rotations as png files, but I noticed I started getting this strange error, "Context leak detected, msgtracer returned -1" when the number of loop iterations increased. For example if I run this code with n = 10, there is no problem. But if I run this code with n=100+, I start getting the "context leak detected" error. 

Minimally reproducible code: 
```python
import pyvista as pv
import numpy as np

# Generate random views of a cube and save render as png
save_path = '/your/folder'
n = 100
cube = pv.Cube()

for i in range(n):
    pl = pv.Plotter(off_screen=True)
    deg_x = np.random.randint(1, 360)
    deg_y = np.random.randint(1, 360)
    deg_z = np.random.randint(1, 360)
    rotated_cube = cube.rotate_x(deg_x, inplace=False)
    rotated_cube.rotate_y(deg_y, inplace=True)
    rotated_cube.rotate_z(deg_z, inplace=True)
    pl.add_mesh(rotated_cube)
    file_name = f'cube_{i}.png'
    file_path = save_path + '/' + file_name
    pl.screenshot(file_path, return_img=False)
    pl.close()
```

I have no idea what this means....and a google search didn't really help. Anyone else experience similar issues, or can anyone else reproduce this error on their machine? 

I'm on a MacBook Pro, 2023 w/ Apple M2 Pro chip if that matters. Happy to provide any additional information as needed.

**Comment by banesullivan** [comment, marked as answer]
You'll likely want to have one plotter instance and update the mesh in that plotter to take many screenshots. There is an upstream issue on the new Apple M1/M2 GPUs leading to this context leak.

See also:

- https://github.com/pyvista/pyvista/issues/3474
- https://gitlab.kitware.com/vtk/vtk/-/issues/18713

Creating many plotters one after another causes the issue. Try one plotter, but moving the mesh in that plotter:

```py
import pyvista as pv
import numpy as np

# Generate random views of a cube and save render as png
save_path = 'foo/'
n = 100
cube = pv.Cube()

pl = pv.Plotter(off_screen=True)
# Use name to replace mesh
pl.add_mesh(cube, name='cube')

for i in range(n):
    deg_x = np.random.randint(1, 360)
    deg_y = np.random.randint(1, 360)
    deg_z = np.random.randint(1, 360)
    rotated_cube = cube.rotate_x(deg_x, inplace=False)
    rotated_cube.rotate_y(deg_y, inplace=True)
    rotated_cube.rotate_z(deg_z, inplace=True)
    
    # Use name to replace mesh
    pl.add_mesh(rotated_cube, name='cube')
    
    file_name = f'cube_{i}.png'
    file_path = save_path + '/' + file_name
    pl.screenshot(file_path, return_img=False)

pl.close()
```

I just ran this on my M2 MacBook without issue.

Also note that updating the mesh in place can be more efficient (rather than replacing the mesh in the scene with the `name` argument)

**Reply by josephko91** [reply]
Thanks so much @banesullivan! That did the trick. 

Thanks for explaining about the context leak issue with the Mac GPUs. I didn't know that was a thing, but I'll keep that in mind.

**Comment by WesleyTheGeolien** [comment]
Is this linked in any means to https://github.com/pyvista/pyvista/discussions/4274 / https://github.com/pyvista/pyvista/issues/2252 ?
